File: programmers/python/입문100/2.3_분수의덧셈★/2-3.py
def solution(numer1, denom1, numer2, denom2):
    # 통분해서 분수더하기
    numer3 = (numer1 * denom2) + (numer2 * denom1)
    denom3 = denom1 * denom2

    # 최대공약수 찾아서 약분하기
    a, b = numer3, denom3
    while b:  # 나머지가 0이 될때까지
        a, b = b, a % b  # 유클리드 호제법
    gcd = a  # 0이 아닌 다른 수가 최대공약수
    numer3, denom3 = numer3 // gcd, denom3 // gcd  # 약분하기

    return [numer3, denom3]  # 배열을 리턴하라고 했는데, 튜플로 리턴해도 문제가 안된다?

File: programmers/cpp/입문100/2.3_분수의덧셈/2-3.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int numer1, int denom1, int numer2, int denom2) {
    vector<int> answer;

    // 코드입력
    // 약분없이 곱셈을 사용하여 분수의 합을 구한다
    int numer = (numer1 * denom2) + (numer2 * denom1);
    int denum = denom1 * denom2;

    // 최대공약수를 구하여 분모분자를 약분한다 (유클리드 호제법)
    int a = numer;
    int b = denum;
    int c;
    
    while(b != 0)
    {
        c = a % b;
        a = b;
        b = c;
    }
    
    int gcd = a;

    // 최대공약수로 약분하기
    numer /= gcd;
    denum /= gcd;
    
    // 벡터에 분자 분모 순서대로 푸쉬백한다        
    answer.push_back(numer);
    answer.push_back(denum);

    return answer;
}
